File: C++17/백준/Silver/2792. 보석 상자/보석 상자.cc
#include <bits/stdc++.h>
using namespace std;
#define MAX 1000000001
int N,M,k,ret,ans;
vector<int> jew;

int check(int mid){
    int cnt=0;
    for (auto j:jew){
        cnt+=(j/mid);
        if (j%mid) cnt++;
    }
    return cnt;
}
int main(){
    cin>>N>>M;
    ans=MAX;
    while (M--){
        cin>>k;
        jew.push_back(k);
    }

    sort(jew.begin(),jew.end());

    int lo=1;int hi=jew[jew.size()-1];
    int mid=(lo+hi)/2;

    while (lo<=hi){
        mid=(lo+hi)/2;
        ret=check(mid);

        if (ret>N) lo=mid+1;
        else{
            hi=mid-1;
            ans=min(ans,mid);
        }
    }
    cout<<ans;
}

File: Java/백준/Silver/2792. 보석 상자/보석 상자.java
import java.util.Scanner;

public class Main {
    static int M,N;
    static int[] ar;
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        N=sc.nextInt(); M=sc.nextInt();
        ar=new int[M];
        int hi=0,mid=0;
        for (int i=0;i<M;i++){
            ar[i]=sc.nextInt();
            hi=Math.max(hi,ar[i]);
        }

        int lo=1,ret=0;

        while (lo<=hi){
            mid=(lo+hi)/2;

            if (check(mid)){//질투심보다 작거나 같다면
                hi=mid-1;
                ret=mid;
            }else{
                lo=mid+1;
            }
        }
        System.out.println(ret);

    }

    static boolean check(int mid){
        int tmp=0;
        for (int i=0;i<M;i++){
            tmp+=ar[i]/mid;
            if (ar[i]%mid!=0) tmp++;
        }

        return tmp<=N;
    }
}
